### PR TITLE
Modifying the LG application identifying object

### PR DESCRIPTION
**Pull request by MahEzzat**
Modifying the LG application identifying object to require the application name, region, locale, and version.

### CODE REVIEW

**Comment by cleemullins** [comment]
@MahEzzat Who would you recommend to do the code review on this?

File: libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/Engine/Pipeline/ResolverPipelineFactory.cs [hunk 1]
```diff
@@ -19,18 +20,45 @@ internal class ResolverPipelineFactory : IResolverPipelineFactory
...
+        }
+        
+        /// <summary>
+        /// Creates a new <see cref="IResolverPipeline"/> object using the passed arguments.
+        /// </summary>
+        /// <param name="endpoint">Language generation runtime api endpoint.</param>
+        /// <param name="endpointKey">Language generation runtime api endpoint key.</param>
+        /// <param name="applicationId">Language generation application id.</param>
+        /// <param name="tokenGenerationEndpoint">Token generation endpoint, default is "https://api.cognitive.microsoft.com/sts/v1.0/issueToken".</param>
+        /// <param name="serviceAgent">A <see cref="IServiceAgent"/> object.</param>
```
Comment (kayousef) [inline libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/Engine/Pipeline/ResolverPipelineFactory.cs]: Update docs

File: libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/Engine/Pipeline/ResolverPipelineFactory.cs [hunk 2]
```diff
@@ -19,18 +20,45 @@ internal class ResolverPipelineFactory : IResolverPipelineFactory
         /// <param name="tokenGenerationEndpoint">Token generation endpoint, default is "https://api.cognitive.microsoft.com/sts/v1.0/issueToken".</param>
         /// <param name="serviceAgent">A <see cref="IServiceAgent"/> object.</param>
         /// <returns>A <see cref="IResolverPipeline"/>.</returns>
-        public IResolverPipeline CreateResolverPipeline(string endpointURI, string endpointKey, string applicationId, string tokenGenerationEndpoint = null, IServiceAgent serviceAgent = null)
+        public IResolverPipeline CreateResolverPipeline(LanguageGenerationApplication languageGenerationApplication, IEndpointProvider endpointProvider)
```
Comment (kayousef) [inline libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/Engine/Pipeline/ResolverPipelineFactory.cs]: Fix docs

File: samples/Basic-Bot/LanguageGenerationUtilities.cs [hunk 3]
```diff
@@ -16,33 +16,21 @@ public static class LanguageGenerationUtilities
...
-                // Set default region to westus
-                endpointRegion = "westus";
-            }
 
-            var languageGenerationEndpoint = $"https://{endpointRegion}.cts.speech.microsoft.com/v1/lg";
-            var tokenIssuingEndpoint = $"https://{endpointRegion}.api.cognitive.microsoft.com/sts/v1.0/issueToken";
+            var application = new LanguageGenerationApplication(applicationId, applicationRegion, applicationLocale, applicationVersion, subscriptionKey);
 
-            var application = new LanguageGenerationApplication(applicationId, endpointKey, languageGenerationEndpoint);
+            var languageGenerationEndpoint = $"https://{applicationRegion}.cts.speech.microsoft.com/v1/lg";
```
Comment (kayousef) [inline samples/Basic-Bot/LanguageGenerationUtilities.cs]: Why do we have this hard coded in multiple regions?

File: samples/Basic-Bot/Startup.cs [hunk 4]
```diff
@@ -123,10 +123,15 @@ public void ConfigureServices(IServiceCollection services)
             var entityStateAccessor = userState.CreateProperty<Dictionary<string, object>>("EntityState");
             services.AddSingleton(entityStateAccessor);
 
+            // The application values must match the ones during creation of the endpoint
             var applicationId = "lgshowcases";
-            var endpointKey = Keys.LanguageGenerationSubscriptionKey;
-            var endpointRegion = "westus"; // The region must be the subscription key's region.
-            var languageGenerationMiddleware = LanguageGenerationUtilities.CreateMiddleware(applicationId, endpointKey, endpointRegion, entityStateAccessor);
+            var applicationRegion = "westus";
```
Comment (kayousef) [inline samples/Basic-Bot/Startup.cs]: Should those be constant somewhere else?

File: samples/Cafe-Bot/MultiTurnPromptsBot.cs [hunk 5]
```diff
@@ -69,10 +69,15 @@ public MultiTurnPromptsBot(BotAccessors accessors)
             _dialogs.Add(new TextPrompt("confirmBooking"));
             _dialogs.Add(new ConfirmPrompt("confirm"));
 
+            // The application values must match the ones during creation of the endpoint
             var applicationId = "cafebot";
-            var endpointKey = Keys.LanguageGenerationSubscriptionKey;
-            var endpointRegion = "westus"; // The region must be the subscription key's region.
-            _languageGenerationResolver = LanguageGenerationUtilities.CreateResolver(applicationId, endpointKey, endpointRegion);
+            var applicationRegion = "westus";
```
Comment (kayousef) [inline samples/Cafe-Bot/MultiTurnPromptsBot.cs]: same

File: tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/LanguageGenerationApplicationTests.cs [hunk 6]
```diff
@@ -7,21 +7,75 @@ namespace Microsoft.Bot.Builder.AI.LanguageGeneration.Tests
     [TestClass]
     public class LanguageGenerationApplicationTests
     {
-        private const string Endpoint = "https://platform.bing.com/speechdx/lg-dev/v1/lg";
+        private readonly string applicationId = "test_application";
+        private readonly string applicationRegion = "westus";
```
Comment (kayousef) [inline tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/LanguageGenerationApplicationTests.cs]: same

File: tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/UtilitiesTests.cs [hunk 7]
```diff
@@ -674,24 +673,24 @@ public void TestActivityModifier_ModifyAttachments_Valid()
...
-            {
-                var dummyAppId = "TEST_ID";
-                var requestBuilder = new RequestBuilder(dummyAppId);
-                requestBuilder.BuildRequest(null, null);
-            }
-            catch (ArgumentNullException e)
-            {
-                Assert.AreEqual("slots", e.ParamName);
-            }
+            var dummyAppId = "TEST_ID";
```
Comment (kayousef) [inline tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/UtilitiesTests.cs]: same

File: samples/Basic-Bot/LanguageGenerationUtilities.cs (hunk 3 above)
Comment (MahEzzat) [reply samples/Basic-Bot/LanguageGenerationUtilities.cs]: This is only placed in two places, here to demonstrate the ability to override the endpoints through the LanguageGenerationOptions object, and in the DefaultRegionalEndpointProvider for the default endpoint provider.

File: samples/Basic-Bot/Startup.cs (hunk 4 above)
Comment (MahEzzat) [reply samples/Basic-Bot/Startup.cs]: No, as these values should be explicitly set by the developer using the samples.

File: samples/Cafe-Bot/MultiTurnPromptsBot.cs (hunk 5 above)
Comment (MahEzzat) [reply samples/Cafe-Bot/MultiTurnPromptsBot.cs]: No, as these values should be explicitly set by the developer using the samples.

File: tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/LanguageGenerationApplicationTests.cs (hunk 6 above)
Comment (MahEzzat) [reply tests/Microsoft.Bot.Builder.AI.LanguageGeneration.Tests/LanguageGenerationApplicationTests.cs]: These are already constants, or you mean something different?

**Pull request merged**